stratum/hal/lib/barefoot/bfrt_table_manager.cc

```diff
@@ -133,69 +70,89 @@ ::util::Status BfrtTableManager::VerifyForwardingPipelineConfig(
-          Uint64ToByteStream(annotation.register_reset_value());
-      min_reset_interval_ms = std::min(annotation.register_reset_interval_ms(),
-                                       min_reset_interval_ms);
-      ::p4::v1::RegisterEntry register_entry;
-      register_entry.set_register_id(reg.preamble().id());
-      register_entry.mutable_data()->set_bitstring(clear_value);
-      register_entry.clear_index();
-      // TODO(max): remove after testing
-      if (mode_ == OPERATION_MODE_SIM) {
-        register_entry.mutable_index()->set_index(0);
-      }
-      ::util::Status status =
-          WriteRegisterEntry(session, ::p4::v1::Update::MODIFY, register_entry);
-      VLOG(1) << "cleared register " << reg.preamble().name() << ".";
-    }
-    APPEND_STATUS_IF_BFRT_ERROR(status, session->endBatch(true));
-    APPEND_STATUS_IF_BFRT_ERROR(status, session->sessionCompleteOperations());
-    auto t2 = absl::Now();
-    CHECK(t1 <= t2);
-    int64 remaining_sleep_ms =
-        std::max(static_cast<int64>(min_reset_interval_ms -
-                                    (t2 - t1) / absl::Milliseconds(1)),
-                 0l);
-    VLOG(1) << "Reset all registers in " << (t2 - t1) / absl::Milliseconds(1)
-            << " milliseconds. Next reset in " << remaining_sleep_ms << "ms.";
-    if (!status.ok()) {
-      LOG(ERROR) << "Error resetting registers: " << status;
+::util::Status BfrtTableManager::SetupRegisterReset(
+    const ::p4::config::v1::P4Info& p4_info) {
+  if (!FLAGS_incompatible_enable_register_reset_annotations) {
+    return ::util::OkStatus();
+  }
+  // Crude check to prevent consecutive pipeline pushes.
+  static bool first_time = true;
+  if (!first_time) {
+    LOG(FATAL) << "Multiple pipeline pushes are not allowed when using "
+                  "register reset annotations.";
+  }
+  first_time = false;
+  if (mode_ == OPERATION_MODE_SIM) {
+    LOG(WARNING)
+        << "Register reset annotations are disabled in simulation mode.";
+    return ::util::OkStatus();
+  }
+
+  // Validate consistent reset intervals.
+  std::vector<uint64> intervals_ms;
+  for (const auto& reg : p4_info.registers()) {
+    ASSIGN_OR_RETURN(
+        P4Annotation annotation,
+        p4_info_manager_->GetSwitchStackAnnotations(reg.preamble().name()));
+    if (annotation.register_reset_interval_ms()) {
+      intervals_ms.push_back(annotation.register_reset_interval_ms());
     }
-    status = ::util::OkStatus();
-    absl::SleepFor(absl::Milliseconds(remaining_sleep_ms));
   }
-  APPEND_STATUS_IF_BFRT_ERROR(status, session->sessionDestroy());
+  if (intervals_ms.empty()) {
+    return ::util::OkStatus();
+  }
+  std::sort(intervals_ms.begin(), intervals_ms.end());
+  auto last = std::unique(intervals_ms.begin(), intervals_ms.end());
+  intervals_ms.erase(last, intervals_ms.end());
+  if (intervals_ms.size() != 1) {
+    RETURN_ERROR(ERR_INVALID_PARAM)
+        << "Inconsistent register reset intervals are not supported.";
+  }
```

Comment:
This logic makes sense to me.
Was wondering if we can simplify to this:

```cpp
auto last = std::unique(intervals_ms.begin(), intervals_ms.end());
if (last != (intervals_ms.begin() + 1)) {
  //return error
}
```
```

Reply:
But we can merge this patch now, I don't really find a big issue.

Reply:
I really like you idea, but have two concerns:

- I don't want to test this change again
- It makes the code in more unreadable and harder to understand